Panic attempting to merge: byte slice is not of expected size

I think this is similar to #10676, which I reported last week and it was fixed in 1.83.7. I'm on 1.84.0, with a new panic which seems related.

Again I have a commit which adds a field into a unique index and also adds data which would not have been accepted under the old version of the index. And i get this panic:

```
❯ dolt merge m959398v30fv1sr7qkjikl45fkoo8cmq
panic: byte slice is not of expected size

goroutine 200 [running]:
github.com/dolthub/dolt/go/store/val.expectSize(...)
        github.com/dolthub/dolt/go/store/val/codec.go:675
github.com/dolthub/dolt/go/store/val.readInt8(...)
        github.com/dolthub/dolt/go/store/val/codec.go:231
github.com/dolthub/dolt/go/store/val.compare({0xf8?, 0x70?}, {0x64b0f48561a1?, 0x1?, 0x0?}, {0x64b0fba1a917, 0x25, 0x1069})
        github.com/dolthub/dolt/go/store/val/tuple_compare.go:102 +0x1644
github.com/dolthub/dolt/go/store/val.(*DefaultTupleComparator).CompareValues(...)
        github.com/dolthub/dolt/go/store/val/tuple_compare.go:70
github.com/dolthub/dolt/go/libraries/doltcore/schema.collationCompare({0x64b0fba1a8b6?, 0x3348d06245d669a?}, {0x0?, 0x97?}, 0xfb59?, {0x64b0f48561a1?, 0x64b0f42f1200?, 0x64b0fb587140?}, {0x64b0fba1a917?, 0x10530ac60?, ...})
        github.com/dolthub/dolt/go/libraries/doltcore/schema/collation_comparator.go:116 +0x98
github.com/dolthub/dolt/go/libraries/doltcore/schema.CollationTupleComparator.Compare({{0x64b0fb58de40?, 0x1052fd828?, 0x64b0fb587258?}}, {0x1097cdde0, 0x64b0fab30dc0}, {0x64b0f4856140, 0x97, 0x97}, {0x64b0fba1a8b6, 0x97, ...}, ...)
        github.com/dolthub/dolt/go/libraries/doltcore/schema/collation_comparator.go:49 +0x278
github.com/dolthub/dolt/go/store/val.(*TupleDesc).Compare(0x623f90d5c15bd697?, {0x1097cdde0?, 0x64b0fab30dc0?}, {0x64b0f4856140?, 0x1097cdde0?, 0x64b0fab30dc0?}, {0x64b0fba1a8b6?, 0x1052190cc?, 0x64b0fb5873b8?})
        github.com/dolthub/dolt/go/store/val/tuple_descriptor.go:186 +0x3c
github.com/dolthub/dolt/go/store/prolly/tree.newLeafCursorAtKey[...]({0x1097cdde0, 0x64b0fab30dc0}, {0x1098662e8, 0x64b0f47e7800}, 0x1044558c8, {0x64b0f4856140?, 0x97, 0x97}, 0x64b0fb5bbda0)
        github.com/dolthub/dolt/go/store/prolly/tree/node_cursor.go:194 +0x108
github.com/dolthub/dolt/go/store/prolly/tree.StaticMap[...].GetPrefix(0x109889900?, {0x1097cdde0?, 0x64b0fab30dc0?}, {0x64b0f4856140?, 0x97, 0x97}, 0x64b0fb5bbda0, 0x64b0fb566980)
        github.com/dolthub/dolt/go/store/prolly/tree/map.go:263 +0x7c
github.com/dolthub/dolt/go/store/prolly/tree.MutableMap[...].GetPrefix(0x109885840, {0x1097cdde0, 0x64b0fab30dc0}, {0x64b0f4856140, 0x97, 0x97}, 0x64b0fb5bbda0, 0x64b0fb566980)
        github.com/dolthub/dolt/go/store/prolly/tree/mutable_map.go:65 +0x104
github.com/dolthub/dolt/go/store/prolly.(*GenericMutableMap[...]).GetPrefix(...)
        github.com/dolthub/dolt/go/store/prolly/tuple_mutable_map.go:145
github.com/dolthub/dolt/go/libraries/doltcore/merge.uniqIndex.findCollisions({{0x1098779a0, 0x64b0f740dd40}, 0x64b0fb5bbe00, 0x64b0fb5bbe60, 0x64b0fb5bbda0, 0x64b0fb5bbc20, {{0x64b0f762d6c0, 0x7, 0x7}, {0x64b0f75a6ab0, ...}}, ...}, ...)
        github.com/dolthub/dolt/go/libraries/doltcore/merge/merge_prolly_rows.go:984 +0x1ec
github.com/dolthub/dolt/go/libraries/doltcore/merge.uniqValidator.validateDiff({{0x1097b3ac0, 0x64b0f5757440}, {0xb2, 0x4a, 0x91, 0xa5, 0x1f, 0x18, 0x1f, 0xf0, ...}, ...}, ...)
        github.com/dolthub/dolt/go/libraries/doltcore/merge/merge_prolly_rows.go:820 +0x438
github.com/dolthub/dolt/go/libraries/doltcore/merge.computeProllyTreePatches(0x64b0fab30dc0, 0x64b0f41a36c0, {0x109874f90, 0x64b0f467db00}, 0x64b0fab63160, 0x64b0fabeac60, {0x58?, 0x4f?, 0x7a?}, {0xf5?, ...}, ...)
        github.com/dolthub/dolt/go/libraries/doltcore/merge/merge_prolly_rows.go:327 +0xd68
github.com/dolthub/dolt/go/libraries/doltcore/merge.mergeProllyTableData.func1()
        github.com/dolthub/dolt/go/libraries/doltcore/merge/merge_prolly_rows.go:453 +0x54
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.19.0/errgroup/errgroup.go:93 +0x4c
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
        golang.org/x/sync@v0.19.0/errgroup/errgroup.go:78 +0x90
```

I will try as a workaround rebasing into 2 separate commits and merging them one at a time

--- a/go/libraries/doltcore/merge/merge_prolly_rows.go
+++ b/go/libraries/doltcore/merge/merge_prolly_rows.go
@@ -746,8 +746,13 @@ func newUniqValidator(ctx *sql.Context, sch schema.Schema, tm *TableMerger, vm *
 	for _, def := range sch.Indexes().AllIndexes() {
 		if !def.IsUnique() {
 			continue
-		} else if !tm.leftSch.Indexes().Contains(def.Name()) {
-			continue // todo: how do we validate in this case?
+		}
+
+		leftDef := tm.leftSch.Indexes().GetByName(def.Name())
+		if leftDef == nil || !leftDef.Equals(def) {
+			// mergeProllySecondaryIndexes rebuilds it from the final merged table and
+			// records any violations there.
+			continue
 		}
 
 		idx, err := indexes.GetIndex(ctx, sch, nil, def.Name())

--- a/go/libraries/doltcore/merge/mutable_secondary_index.go
+++ b/go/libraries/doltcore/merge/mutable_secondary_index.go
@@ -47,9 +47,10 @@ func GetMutableSecondaryIdxs(ctx *sql.Context, ourSch, sch schema.Schema, tableN
 	return mods, nil
 }
 
-// GetMutableSecondaryIdxsWithPending returns a MutableSecondaryIdx for each secondary index in |indexes|. If an index
-// is listed in the given |sch|, but does not exist in the given |indexes|, then it is skipped. This is useful when
-// merging a schema that has a new index, but the index does not exist on the index set being modified.
+// GetMutableSecondaryIdxsWithPending returns a [MutableSecondaryIdx] for each secondary index in |indexes|
+// whose definition in |ourSch| matches the corresponding definition in |sch|. Indexes present in |sch| but
+// absent from |indexes|, or whose definition has changed between |ourSch| and |sch|, are skipped. The
+// caller is responsible for rebuilding those indexes from scratch after the merge completes.
 func GetMutableSecondaryIdxsWithPending(ctx *sql.Context, ns tree.NodeStore, ourSch, sch schema.Schema, tableName string, indexes durable.IndexSet, pendingSize int) ([]MutableSecondaryIdx, error) {
 	mods := make([]MutableSecondaryIdx, 0, sch.Indexes().Count())
 	for _, index := range sch.Indexes().AllIndexes() {
@@ -80,11 +81,15 @@ func GetMutableSecondaryIdxsWithPending(ctx *sql.Context, ns tree.NodeStore, our
 		// TODO: This isn't technically required, but correctly handling updating secondary indexes when only some
 		// of the table's rows have been updated is difficult to get right.
 		// Dropping the index is potentially slower but guaranteed to be correct.
+		ourIndex := ourSch.Indexes().GetByName(index.Name())
+		if ourIndex == nil || !ourIndex.Equals(index) {
+			continue
+		}
+
 		idxKeyDesc := m.KeyDesc()
 		if schema.IsKeyless(sch) {
 			idxKeyDesc = idxKeyDesc.PrefixDesc(idxKeyDesc.Count() - 1)
 		}
-
 		if !idxKeyDesc.Equals(index.Schema().GetKeyDescriptor(ns)) {
 			continue
 		}

--- a/go/libraries/doltcore/merge/mutable_secondary_index.go
+++ b/go/libraries/doltcore/merge/mutable_secondary_index.go
@@ -68,6 +68,11 @@ func GetMutableSecondaryIdxsWithPending(ctx *sql.Context, ns tree.NodeStore, our
 			continue
 		}
 
+		ourIndex := ourSch.Indexes().GetByName(index.Name())
+		if ourIndex == nil || !ourIndex.Equals(index) {
+			continue
+		}
+
 		idx, err := indexes.GetIndex(ctx, sch, nil, index.Name())
 		if err != nil {
 			return nil, err
@@ -81,15 +86,11 @@ func GetMutableSecondaryIdxsWithPending(ctx *sql.Context, ns tree.NodeStore, our
 		// TODO: This isn't technically required, but correctly handling updating secondary indexes when only some
 		// of the table's rows have been updated is difficult to get right.
 		// Dropping the index is potentially slower but guaranteed to be correct.
-		ourIndex := ourSch.Indexes().GetByName(index.Name())
-		if ourIndex == nil || !ourIndex.Equals(index) {
-			continue
-		}
-
 		idxKeyDesc := m.KeyDesc()
 		if schema.IsKeyless(sch) {
 			idxKeyDesc = idxKeyDesc.PrefixDesc(idxKeyDesc.Count() - 1)
 		}
+
 		if !idxKeyDesc.Equals(index.Schema().GetKeyDescriptor(ns)) {
 			continue
 		}

--- a/go/libraries/doltcore/sqle/enginetest/dolt_queries_schema_merge.go
+++ b/go/libraries/doltcore/sqle/enginetest/dolt_queries_schema_merge.go
@@ -1428,6 +1428,98 @@ var SchemaChangeTestsConstraints = []MergeScriptTest{
 			},
 		},
 	},
+	{
+		// https://github.com/dolthub/dolt/issues/10753.
+		Name: "adding a column to a unique index where merged rows violate the new index",
+		AncSetUpScript: []string{
+			"set autocommit = 0;",
+			"CREATE TABLE t (pk INT PRIMARY KEY, a INT, b TINYINT, UNIQUE KEY uq (a));",
+			"INSERT INTO t VALUES (1, 10, 1);",
+		},
+		RightSetUpScript: []string{
+			"ALTER TABLE t DROP INDEX uq;",
+			"ALTER TABLE t ADD UNIQUE KEY uq (a, b);",
+			"INSERT INTO t VALUES (2, 20, 2);",
+		},
+		LeftSetUpScript: []string{
+			"INSERT INTO t VALUES (3, 20, 2);",
+		},
+		Assertions: []queries.ScriptTestAssertion{
+			{
+				Query:    "call dolt_merge('right');",
+				Expected: []sql.Row{{"", 0, 1, "conflicts found"}},
+			},
+			{
+				Query:    "select * from dolt_conflicts;",
+				Expected: []sql.Row{},
+			},
+			{
+				Query:    "select * from dolt_constraint_violations;",
+				Expected: []sql.Row{{"t", uint(2)}},
+			},
+			{
+				Query: "select violation_type, pk, a, b, violation_info from dolt_constraint_violations_t order by pk;",
+				Expected: []sql.Row{
+					{"unique index", 2, 20, int8(2), merge.UniqCVMeta{Columns: []string{"a", "b"}, Name: "uq"}},
+					{"unique index", 3, 20, int8(2), merge.UniqCVMeta{Columns: []string{"a", "b"}, Name: "uq"}},
+				},
+			},
+			{
+				Query:    "select * from dolt_status;",
+				Expected: []sql.Row{{"t", byte(0), "constraint violation"}},
+			},
+			{
+				Query:    "UPDATE t SET a = 30 WHERE pk = 3;",
+				Expected: []sql.Row{{types.OkResult{RowsAffected: 1, Info: plan.UpdateInfo{Matched: 1, Updated: 1}}}},
+			},
+			{
+				Query:    "DELETE FROM dolt_constraint_violations_t;",
+				Expected: []sql.Row{{types.NewOkResult(2)}},
+			},
+			{
+				Query:    "call dolt_commit('-am', 'merge right, resolve unique index violation');",
+				Expected: []sql.Row{{doltCommit}},
+			},
+			{
+				Query: "SELECT * FROM t ORDER BY pk;",
+				Expected: []sql.Row{
+					{1, 10, int8(1)},
+					{2, 20, int8(2)},
+					{3, 30, int8(2)},
+				},
+			},
+		},
+	},
+	{
+		// https://github.com/dolthub/dolt/issues/10753.
+		Name: "adding a column to a unique index and inserting data valid only under the new index",
+		AncSetUpScript: []string{
+			"CREATE TABLE t (pk INT PRIMARY KEY, a INT, b TINYINT, UNIQUE KEY uq (a));",
+			"INSERT INTO t VALUES (1, 10, 1);",
+		},
+		RightSetUpScript: []string{
+			"ALTER TABLE t DROP INDEX uq;",
+			"ALTER TABLE t ADD UNIQUE KEY uq (a, b);",
+			"INSERT INTO t VALUES (2, 10, 2);",
+		},
+		LeftSetUpScript: []string{
+			"INSERT INTO t VALUES (3, 30, 3);",
+		},
+		Assertions: []queries.ScriptTestAssertion{
+			{
+				Query:    "call dolt_merge('right');",
+				Expected: []sql.Row{{doltCommit, 0, 0, "merge successful"}},
+			},
+			{
+				Query: "SELECT * FROM t ORDER BY pk;",
+				Expected: []sql.Row{
+					{1, 10, int8(1)},
+					{2, 10, int8(2)},
+					{3, 30, int8(3)},
+				},
+			},
+		},
+	},
 }
 
 // SchemaChangeTestsTypeChanges holds test scripts for schema merge where column types have changed. Note that

